### ISSUE TITLE
support date ranges in query syntax

### ISSUE THREAD

**Issue by mschoch** [blevesearch/bleve#53]

**Comment by zaccone** [comment]
Looks like it's a matter of parsing input text and trying to figure provided input compatible with either iso8601 or rfc3339 formats execute NewDateRangeQuery().
Syntax would look similar o what bleve has today:

<field>:2015-01-05
<field>: > 2015-01-05
<field>: < 2015-01-05 
<field>: 2014-01-01 TO 2015-01-01

I think the only change required is modification of nex and yacc files so the grammar is extended. The juicy mean seems to be already implemented in bleve (query_date_range.go file).
What are your opinions on proposed syntax? Maybe there is something standardized and better?

**Comment by zaccone** [comment]
Looks like it's a matter of parsing input text and trying to figure provided input compatible with either iso8601 or rfc3339 formats execute NewDateRangeQuery().
Syntax would look similar o what bleve has today:

field:2015-01-05
field: > 2015-01-05
field: < 2015-01-05 
field: 2014-01-01 TO 2015-01-01

I think the only change required is modification of nex and yacc files so the grammar is extended. The juicy mean seems to be already implemented in bleve (query_date_range.go file).
What are your opinions on proposed syntax? Maybe there is something standardized and better?

**Comment by mschoch** [comment]
There is some further discussion on the topic here:

https://github.com/blevesearch/bleve/issues/151

One user started work to rewrite the entire query parsing, but I haven't heard any update on this effort.

The Lucene query syntax has pretty generic support for "range queries:

field:x

exact match, x could be text term, number or date

field:>x

greater than, probably only useful for number and date (meaningful for term, but perhaps too costly to allow by default)

Same for >,<,>=,<=

Additionally, we should support the TO operator as the preferred way to perform queries with bounds on both sides.

So, I think to do it right, its a bit more work than just hacking in support for dates, but I don't think we necessarily have to rewrite the whole query processing.

**Comment by zaccone** [comment]
I agree >, >=, <, <= should be supporting both dates and numbers (at least for starters) - do you think those should be different rules in grammar or bleve should distinguish between number and date and apply different functions, NewDateRangeQuery() for dates NewNumericRangeQuery() for number and thie Inclusive counterparts? 
If possible, I vote for splitting use cases at the grammar level.

**Comment by zaccone** [comment]
I have created a pull request regaring this issue: https://github.com/blevesearch/bleve/pull/249

In the end I added a support for operators <, <=, >, >= for strings and strings only. The main reason behind is that we need to parse phrases, not terms, as date usually consists whitespaces, ':' and '-' characters.

What needs to be done in the PR is adding the remiaining test, so all date time formats are being checked, and implement start TO stop  range search.
